video/out/w32_common.c

```diff
@@ -795,33 +795,46 @@ static void update_window_style(struct vo_w32_state *w32)
     w32->windowrc = wr;
 }
 
-// Adjust rc size and position if its size is larger than rc2.
+// Adjust rc size and/or position so that it fits inside rc2.
 // returns true if the rectangle was modified.
 static bool fit_rect(RECT *rc, RECT *rc2)
 {
     // Calculate old size and maximum new size
     int o_w = rect_w(*rc), o_h = rect_h(*rc);
     int n_w = rect_w(*rc2), n_h = rect_h(*rc2);
-    if (o_w <= n_w && o_h <= n_h)
-        return false;
 
-    // Apply letterboxing
-    const float o_asp = o_w / (float)MPMAX(o_h, 1);
-    const float n_asp = n_w / (float)MPMAX(n_h, 1);
-    if (o_asp > n_asp) {
-        n_h = n_w / o_asp;
+    if (o_w <= n_w && o_h <= n_h) {
+        if (rc->top >= rc2->top && rc->left >= rc2->left &&
+            rc->bottom <= rc2->bottom && rc->right <= rc2->right)
+        {
+            return false;
+        }
+        // size is ok, but rc is not positioned [fully] inside rc2
+        n_w = o_w;
+        n_h = o_h;
+
     } else {
-        n_w = n_h * o_asp;
+        // Apply letterboxing - rc size if too big
+        const float o_asp = o_w / (float)MPMAX(o_h, 1);
+        const float n_asp = n_w / (float)MPMAX(n_h, 1);
+        if (o_asp > n_asp) {
+            n_h = n_w / o_asp;
+        } else {
+            n_w = n_h * o_asp;
+        }
     }
 
-    // Calculate new position and save the rect
-    const int x = rc->left + o_w / 2 - n_w / 2;
-    const int y = rc->top + o_h / 2 - n_h / 2;
+    // Calculate new position with same center as before
+    int x = rc->left + o_w / 2 - n_w / 2;
+    int y = rc->top + o_h / 2 - n_h / 2;
+    // If we're not entirely inside rc2 - move just enough to make it inside
+    x = x < rc2->left ? rc2->left : x + n_w > rc2->right ? rc2->right - n_w : x;
+    y = y < rc2->top ? rc2->top : y + n_h > rc2->bottom ? rc2->bottom - n_h : y;
```

Comment:
LGTM, except these lines are really hard to follow.

The original code didn't take into account the difference in rect positions, indeed.

Reply:
Thanks, also for clarifying. I'll try to refactor these lines.

My main concern, however, is the big comment below with FIXME. Is it reasonably safe to assume that the fit code is not reached after fullscreen/maximized/minimized/sleep/etc?

It was too hard for me to follow the states, but I'll take your word if you feel it's reasonably safe (I'd prefer to understand though, if you care to explain).

Reply:
I replied in the comment below about the conditions when `fit_window_on_screen` is called.